src/gui/log/logmodel.cpp

```diff
@@ -0,0 +1,179 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "logmodel.h"
+
+#include <QApplication>
+#include <QDateTime>
+#include <QColor>
+#include <QPalette>
+
+#include "base/logger.h"
+
+BaseLogModel::BaseLogModel(int initialSize, QObject *parent)
+    : QAbstractListModel(parent)
+    , m_msgCount(initialSize)
+    , m_cache(64)
+{
+    if (m_msgCount == 0)
+        m_startIndex = 0;
+}
+
+int BaseLogModel::rowCount(const QModelIndex &) const
+{
+    if (m_startIndex == -1)
+        return 0;
+    return m_msgCount;
+}
+
+int BaseLogModel::columnCount(const QModelIndex &) const
+{
+    return 1;
+}
+
+QVariant BaseLogModel::data(const QModelIndex &index, const int role) const
+{
+    if (!index.isValid() || (m_msgCount == 0))
+        return {};
+
+    const int i = m_msgCount - index.row() + m_startIndex - 1;
+
+    Item *item = m_cache[i];
+    if (!item) {
+        item = new Item{rowData(i)};
+        m_cache.insert(i, item, 1);
+    }
+
+    switch (role) {
+    case Qt::DisplayRole:
+        return item->displayRole;
+    case Qt::ForegroundRole:
+        return item->foregroundRole;
+    case Qt::UserRole:
+        return item->userRole;
+    default:
+        return {};
+    }
+}
+
+void BaseLogModel::addNewMessage(int index)
+{
+    if (m_startIndex == -1)
+        m_startIndex = index;
+
+    beginInsertRows(QModelIndex(), 0, 0);
+    ++m_msgCount;
+    endInsertRows();
+
+    const int count = rowCount();
+    if (count > MAX_LOG_MESSAGES) {
+        const int lastMessage = count - 1;
+        beginRemoveRows(QModelIndex(), lastMessage, lastMessage);
+        --m_msgCount;
+        endRemoveRows();
+    }
+}
+
+void BaseLogModel::reset()
+{
+    beginResetModel();
+    m_startIndex = m_msgCount;
+    m_msgCount = 0;
+    endResetModel();
+}
+
+LogMessageModel::LogMessageModel(QObject *parent) : 
+    BaseLogModel(Logger::instance()->messageCount(), parent)
+{
+    connect(Logger::instance(), &Logger::newLogMessage, this, &LogMessageModel::handleNewMessage);
+}
+
+void LogMessageModel::handleNewMessage(const Log::Msg &msg)
+{
+    addNewMessage(msg.id);
+}
+
+BaseLogModel::Item LogMessageModel::rowData(const int index) const
+{
+    const Log::Msg msg = Logger::instance()->message(index);
+
```

Comment:
move this empty line below `item` declaration.

Reply:
done.